File: app/src/main/java/com/joeshuff/dddungeongenerator/NewDungeonActivity.java
package com.joeshuff.dddungeongenerator;

import android.content.Intent;
import android.support.design.button.MaterialButton;
import android.support.design.widget.TextInputEditText;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.AppCompatSeekBar;
import android.support.v7.widget.SwitchCompat;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewTreeObserver;
import android.widget.*;
import com.google.gson.Gson;
import com.joeshuff.dddungeongenerator.memory.MemoryController;
import com.joeshuff.dddungeongenerator.memory.MemoryGeneration;
import com.warkiz.tickseekbar.OnSeekChangeListener;
import com.warkiz.tickseekbar.SeekParams;
import com.warkiz.tickseekbar.TickSeekBar;
import org.json.JSONException;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public class NewDungeonActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_new);

        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setTitle("Configure Dungeon");

        loadUI();
        initialiseControls();
    }

    TextInputEditText dungeonSeed;

    AppCompatSeekBar roomSize;

    ScrollView scrollingContent;

    SwitchCompat longerCorridorSwitch;
    TextView longCorridorStatus;

    LinearLayout scrollDownView;

    TickSeekBar monsterFrequency;
    TickSeekBar treasureFrequency;
    TickSeekBar trapFrequency;
    TickSeekBar dungeonDepth;

    MaterialButton generateButton;

    private void loadUI() {
        dungeonSeed = findViewById(R.id.seedInput);

        roomSize = findViewById(R.id.roomSizeBar);

        longerCorridorSwitch = findViewById(R.id.longerCorridorsSwitch);
        longCorridorStatus = findViewById(R.id.longerCorridorsStatus);

        scrollingContent = findViewById(R.id.scrollContent);
        scrollDownView = findViewById(R.id.scrollDownPrompt);

        monsterFrequency = findViewById(R.id.monster_frequency);
        treasureFrequency = findViewById(R.id.treasure_frequency);
        trapFrequency = findViewById(R.id.trap_frequency);
        dungeonDepth = findViewById(R.id.dungeon_depth);

        generateButton = findViewById(R.id.generateButton);
    }

    private void initialiseControls() {
        scrollingContent.setOnScrollChangeListener((v, scrollX, scrollY, oldScrollX, oldScrollY) -> {
            View view = (View) scrollingContent.getChildAt(scrollingContent.getChildCount()-1);
            int diff = (view.getBottom()-(scrollingContent.getHeight()+scrollY));

            if (diff == 0) {
                //HIDE TEXT
                scrollDownView.animate().alpha(0f).setDuration(200).start();
            }
            else {
                scrollDownView.animate().alpha(1f).setDuration(200).start();
            }
        });

        scrollingContent.getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
            @Override
            public void onGlobalLayout() {
                int viewHeight = scrollingContent.getMeasuredHeight();
                int contentHeight = scrollingContent.getChildAt(0).getHeight();
                if(viewHeight - contentHeight < 0) {
                    // scrollable
                } else {
                    scrollDownView.animate().alpha(0).start();
                }
                scrollingContent.getViewTreeObserver().removeOnGlobalLayoutListener(this);
            }
        });

        longerCorridorSwitch.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked) {
                longCorridorStatus.setText("The longest corridors will be prioritised");
            } else {
                longCorridorStatus.setText("The shortest corridors will be prioritised");
            }
        });

        generateButton.setOnClickListener(e -> genButtonClicked());
    }

    private String generateRandomDungeonName() {
        return "Awesome Dungeon Name";
    }

    private void genButtonClicked() {
        Intent generator = new Intent(this, GeneratingActivity.class);

        String seed = dungeonSeed.getText().toString();

        if (seed.isEmpty()) {
            seed = System.currentTimeMillis() + "";

            List<Character> list = seed.chars().mapToObj(c -> new Character((char) c)).collect(Collectors.toList());
            Collections.shuffle(list);
            StringBuilder sb = new StringBuilder();
            list.forEach(c -> sb.append(c));

            seed = sb.toString();
        }

        FirebaseTracker.EVENT(this, "DungeonCreate", "CreatedWithSeed:" + seed);

        MemoryGeneration memory = new MemoryGeneration(seed,
                roomSize.getProgress(),
                longerCorridorSwitch.isChecked(),
                0.5 + ((monsterFrequency.getProgressFloat() / 100.0f) * 2.5f),
                0.5 + ((trapFrequency.getProgressFloat() / 100.0f) * 2.5f),
                0.5 + ((treasureFrequency.getProgressFloat() / 100.0f) * 2.5f),
                1 + (Math.abs(dungeonDepth.getProgress() - 50) / 50) * 2,
                dungeonDepth.getProgress()
        );

        try {
            MemoryController.addToMemory(getApplicationContext(), memory);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        generator.putExtra("instructions", new Gson().toJson(memory));

        startActivity(generator);
        finish();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                supportFinishAfterTransition();
                return true;
        }

        return false;
    }
}

File: app/src/main/java/com/joeshuff/dddungeongenerator/screens/create/NewDungeonActivity.kt
package com.joeshuff.dddungeongenerator.screens.create

import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.view.ViewTreeObserver.OnGlobalLayoutListener
import android.widget.CompoundButton
import android.widget.LinearLayout
import android.widget.ScrollView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.AppCompatSeekBar
import androidx.appcompat.widget.SwitchCompat
import com.google.android.material.button.MaterialButton
import com.google.android.material.textfield.TextInputEditText
import com.google.gson.Gson
import com.joeshuff.dddungeongenerator.R
import com.joeshuff.dddungeongenerator.memory.MemoryController
import com.joeshuff.dddungeongenerator.memory.MemoryGeneration
import com.joeshuff.dddungeongenerator.util.FirebaseTracker
import com.warkiz.tickseekbar.TickSeekBar
import kotlinx.android.synthetic.main.activity_new.*
import org.json.JSONException
import java.util.*
import java.util.function.Consumer
import java.util.stream.Collectors

class NewDungeonActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_new)

        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.title = "Configure Dungeon"

        initialiseControls()
    }

    private fun initialiseControls() {
        scrollContent.setOnScrollChangeListener { v: View?, scrollX: Int, scrollY: Int, oldScrollX: Int, oldScrollY: Int ->
            val view = scrollContent.getChildAt(scrollContent.childCount - 1) as View
            val diff = view.bottom - (scrollContent.height + scrollY)
            if (diff == 0) {
                //HIDE TEXT
                scrollDownPrompt.animate().alpha(0f).setDuration(200).start()
            } else {
                scrollDownPrompt.animate().alpha(1f).setDuration(200).start()
            }
        }

        scrollContent.viewTreeObserver.addOnGlobalLayoutListener(object : OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                val viewHeight = scrollContent.measuredHeight
                val contentHeight = scrollContent.getChildAt(0).height
                if (viewHeight - contentHeight < 0) {
                    // scrollable
                } else {
                    scrollDownPrompt.animate().alpha(0f).start()
                }
                scrollContent.viewTreeObserver.removeOnGlobalLayoutListener(this)
            }
        })

        longerCorridorsSwitch.setOnCheckedChangeListener { buttonView: CompoundButton?, isChecked: Boolean ->
            if (isChecked) {
                longerCorridorsStatus.text = "The longest corridors will be prioritised"
            } else {
                longerCorridorsStatus.text = "The shortest corridors will be prioritised"
            }
        }

        generateButton.setOnClickListener { genButtonClicked() }
    }

    private fun genButtonClicked() {
        val generator = Intent(this, GeneratingActivity::class.java)
        var seed = seedInput.text.toString()

        if (seed.isEmpty()) {
            seed = System.currentTimeMillis().toString() + ""
            val list = seed.chars().mapToObj { c: Int -> c.toChar() }.collect(Collectors.toList())
            list.shuffle()
            val sb = StringBuilder()
            list.forEach(Consumer { c: Char? -> sb.append(c) })
            seed = sb.toString()
        }

        FirebaseTracker.EVENT(this, "DungeonCreate", "CreatedWithSeed:$seed")
        val memory = MemoryGeneration(seed,
                roomSizeBar.progress,
                longerCorridorsSwitch.isChecked,
                0.5 + monster_frequency.progressFloat / 100.0f * 2.5f,
                0.5 + trap_frequency.progressFloat / 100.0f * 2.5f,
                0.5 + treasure_frequency.progressFloat / 100.0f * 2.5f,
                (1 + Math.abs(dungeon_depth.progress - 50) / 50 * 2).toDouble(),
                dungeon_depth.progress
        )

        try {
            MemoryController.addToMemory(applicationContext, memory)
        } catch (e: JSONException) {
            e.printStackTrace()
        }

        generator.putExtra("instructions", Gson().toJson(memory))
        startActivity(generator)
        finish()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                supportFinishAfterTransition()
                return true
            }
        }
        return false
    }
}